Expose entire head pod Service to the user

architkulkarni opened the pull request:
<!-- Thank you for your contribution! -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
This PR exposes the head pod service and all of its fields as part of the RayCluster CRD.  This allows the user to specify labels, annotations, and other fields such as `LoadBalancerIP` for the head service.

Here's the definition of the priorities for when there's overlap (evolved from discussion below). In summary, the new HeadPodService fields are lower priority.

Note that there are three places where labels are defined in the code:
  - (1) There's the `labels` arg passed in to `BuildServiceForHeadPod`. These originate from `HeadGroupSpec.Template.ObjectMeta.Labels`.
  - (2) There's the `default_labels`(the five default labels)
  - (3) There's user-specified custom labels in `HeadService`.

Here's the definitions of the priorities:
- Name: If it's specified in `HeadService`, override the default.
- Namespace: *ignore* user-provided namespace in `HeadService` because we must enforce it to be the same namespace as the RayCluster.
- `annotations` field: The priority is `Spec.HeadServiceAnnotations` > custom `HeadService` annotations.
- `Selector` field: Only use the keys from (2), don't add any new keys. Just update the values of the keys from (2) with the values from (1).  Always ignore keys and values from (3).
- `Labels:` field: Equal to `Selector:` + (3).  In case of overlaps, values from (3) should have lowest priority.

- `Ports` field: There is already some logic in `getPortsFromCluster` to add default ports only if `HeadGroupSpec.Template.Spec.Containers[index].Ports` is empty.  In this PR, we keep this logic and simply append all the ports from `HeadService`.  However, we can consider more complicated merging behavior in the future (or in this PR if necessary), like merging based on the key `port`.
- `Type`: If it's specified in `HeadService`, override the default.

~TODO: Add unit test for the selector merging behavior~
~TODO: Add test in raycluster controller to execute control plane logic. Example: #1037~

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
Closes https://github.com/ray-project/kuberay/issues/877
<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

kevin85421 commented on ray-operator/controllers/ray/raycluster_controller_fake_test.go:
```diff
@@ -1180,3 +1181,58 @@ func TestUpdateStatusObservedGeneration(t *testing.T) {
 	assert.Equal(t, cluster.ObjectMeta.Generation, cluster.Status.ObservedGeneration)
 }
+
+func TestUserSpecifiedHeadService(t *testing.T) {
```
Fake tests are meant for testing the control plane of Kubernetes, but it seems like this particular test does not launch the control plane. Maybe we can move the test to `service_test.go`. You can refer to #874 for more details.

architkulkarni replied:
Thanks for the pointer! Moved

kevin85421 commented on ray-operator/controllers/ray/raycluster_controller_fake_test.go:
```diff
@@ -48,21 +48,22 @@ import (
+	testPods                      []runtime.Object
+	testPodsNoHeadIP              []runtime.Object
+	testRayCluster                *rayiov1alpha1.RayCluster
+	testRayClusterWithHeadService *rayiov1alpha1.RayCluster
```
Github complains that "var `testRayClusterWithHeadService` is unused (unused)"

architkulkarni replied:
Should be fixed now

kevin85421 commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -40,6 +40,42 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
+	if cluster.Spec.HeadGroupSpec.HeadService != nil {
+		// Use the provided HeadService
+		// Priority: HeadService > HeadGroupSpec.PodTemplateSpec > default labels
+		for k, v := range labels {
```
You can take a look at #874 for more context about label priority.

architkulkarni replied:
Thanks for the pointer -- can you help summarize the conclusion from that discussion as it relates to the current PR? I don't have much expertise here so I'd opt to just follow your recommendation.

Is the conclusion the following?
- Fill the `Selector` field with labels, according to the priority `HeadGroupSpec.PodTemplateSpec > default labels > HeadService`
- Fill the `Labels` field with `HeadService` labels only

kevin85421 replied:
For selector, we should always honor `HeadGroupSpec.PodTemplateSpec`. If a mismatch exists between head Pod's label (`HeadGroupSpec.PodTemplateSpec`) and head service's selector, workers cannot talk to the head Pod anymore.

* Head Service:
  * selector: `default labels` (if overlaps exist, respect `HeadGroupSpec.PodTemplateSpec` and ignore `HeadService`.) 
    * We will only have 5 key-value pairs (the number of default labels) here.
  * labels: selector + `HeadService` (if overlaps exist, respect selector and ignore `HeadService`)
    * We may have more than 5 key-value pairs here. 

This is a bit different from the section ["My current thought"](https://github.com/ray-project/kuberay/pull/874#pullrequestreview-1264771952) because some users may use labels to find the head service.

kevin85421 commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -40,6 +40,42 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
 
+	defaultAppProtocol := DefaultServiceAppProtocol
+	ports := getServicePorts(cluster)
+	if cluster.Spec.HeadGroupSpec.HeadService != nil {
```
We may need to consider using deep copy instead. See these links ([link1](https://alenkacz.medium.com/shallow-copy-of-a-go-struct-in-golang-is-never-a-good-idea-83be60106af8), [link2](https://echorand.me/posts/go-values-references-etc/)) for more details.

architkulkarni replied:
(Fixed)

kevin85421 commented on helm-chart/ray-cluster/values.yaml:
```diff
@@ -216,3 +216,4 @@ additionalWorkerGroups:
 service:
   # This is optional, and the default is ClusterIP.
   type: ClusterIP
+  headService: {}
```
Add a comment and add this link: https://pkg.go.dev/k8s.io/Kubernetes/pkg/api/v1#Service

architkulkarni replied:
Makes sense, thanks for the link. Done

kevin85421 commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -40,6 +40,42 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
+			}
+		}
+
+		return cluster.Spec.HeadGroupSpec.HeadService, nil
```
We cannot return here. 

1. `Name` in ObjectMeta is an optional field (see [here](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#ObjectMeta)). If users do not specify the name of the service in `HeadService`, we should generate a deterministic name for them.

2. `Namespace` is also an optional field. If we did not specify the namespace, the Service will be created in the 
"default" namespace. We should enforce it to be in the same namespace as RayCluster.

architkulkarni replied:
Oh wow good catch, that's a bad oversight.  Updated

kevin85421 reviewed (commented):
We should have a clear story for every field in:

```go
	service := &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:        utils.GenerateServiceName(cluster.Name),
			Namespace:   cluster.Namespace,
			Labels:      labels,
			Annotations: annotations,
		},
		Spec: corev1.ServiceSpec{
			Selector: labels,
			Ports:    []corev1.ServicePort{},
			Type:     cluster.Spec.HeadGroupSpec.ServiceType,
		},
	}
```

After you address the comments, we should discuss further about the tests. If this PR has bugs, the Pods may not be able to talk to each other.

architkulkarni commented:
@kevin85421 I've addressed all the comments and I think I'm ready for further discussion about the tests and the merging behavior for when there's overlaps.
- Name: If it's specified in `HeadServiceSpec`, override the default.
- Namespace: If it's specified in `HeadServiceSpec`, override the default.
- `labels` field: This is complicated. 
  - (1) There's the `labels` arg passed in to `BuildServiceForHeadPod`. These originate from `HeadGroupSpec.Template.ObjectMeta.Labels`.
  - (2) There's the `default_labels`(the five default labels)
  - (3) There's user-specified custom labels in `HeadServiceSpec`.
  - For this part, prior to this PR the priority was (1) > (2).  After this PR, the priority is (1) > (2) > (3)
- `annotations` field: The priority is `Spec.HeadServiceAnnotations` > custom `HeadServiceSpec` annotations.
- `Selector` field: Completely ignore the `HeadServiceSpec` selector and labels. Only use (1) and (2), with priority (1) > (2) as before.
- `Ports` field: There is already some logic in `getPortsFromCluster` to add default ports only if `HeadGroupSpec.Template.Spec.Containers[index].Ports` is empty.  In this PR, we keep this logic and simply append all the ports from `HeadServiceSpec`.  However, we can consider more complicated merging behavior in the future (or in this PR if necessary), like merging based on the key `port`.
- `Type`: If it's specified in `HeadServiceSpec`, override the default.

architkulkarni commented:
Btw, is "Compatibility test - nightly" known to be flaky? It's failed a few times, but it also seems to pass once in a while.

kevin85421 commented:
> Btw, is "Compatibility test - nightly" known to be flaky? It's failed a few times, but it also seems to pass once in a while.

All compatibility tests except for 1.13.0 are flaky, but they can usually pass after multiple attempts.

kevin85421 commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -40,27 +41,72 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
+		}
+
+		// If the user has not specified a namespace, use the cluster's namespace
+		if headService.ObjectMeta.Namespace == "" {
```
Can workers connect to the head Pod via the head service if they specify namespace which is different from the RayCluster's namespace?

architkulkarni replied:
Discussed offline -- we will ignore the user specified namespace and raise a warning.

architkulkarni replied:
Addressed

kevin85421 commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -40,27 +41,72 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
-			Namespace:   cluster.Namespace,
+			Name:        default_name,
+			Namespace:   default_namespace,
 			Labels:      labels,
```
`Labels` and `Selector` are the same here. Will this cause connection issues for worker/head Pods if users update the value of custom labels for the head service?

architkulkarni replied:
Discussed offline. Conclusion for Selector: Only use the keys from (2), don't add any new keys. Just update the values of the keys from (2) with the values from (1) if (1) has those keys. Always ignore keys and values from (3))

kevin85421 commented:
"Compatibility test - nightly" starts failing consistently yesterday. I am not sure whether it is related to Ray 2.4.0 release or not.

architkulkarni commented:
@kevin85421 I believe all comments have been addressed and I've added new tests in `raycluster_controller_test`. I've updated the finalized merging behavior in the PR description.  Ready for another review, thanks!

kevin85421 reviewed (commented):
The iteration order of a Map in Go is not guaranteed to be the same from one iteration to the next. Additionally, we may need to consider whether we need to sort slices that are created by copying key-value pairs from a map. Example issue: https://github.com/ray-project/kuberay/pull/1056

architkulkarni commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -30,38 +30,93 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
 		}
 	}
 
+	labels_for_service := selector
```
Ahh, thanks for the reminder (again...)

architkulkarni replied:
Fixed

architkulkarni commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -30,38 +30,93 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
+		headService.ObjectMeta.Namespace = default_namespace
+
+		// If the user has not specified a service type, use the cluster's service type
+		if headService.Spec.Type == "" {
```
Done

architkulkarni commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -30,38 +30,93 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
+
+		// If the user has specified a namespace, ignore it and raise a warning
+		if headService.ObjectMeta.Namespace != "" && headService.ObjectMeta.Namespace != default_namespace {
+			fmt.Printf("Warning: Ignoring namespace %s provided in HeadGroupSpec.HeadService "+
```
Done, using an existing `log` object from the RayCluster-Controller.

architkulkarni commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -30,38 +30,93 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
+		// For the selector, ignore any custom HeadService selectors or labels.
+		headService.Spec.Selector = selector
+
+		// Merge annotations with custom HeadService annotations. If there are overlaps,
```
Done, added a log for each time a key is overwritten.

architkulkarni commented:
Tests are failing, I need to fix the log syntax.

kevin85421 reviewed (commented):
The implementation looks good to me. 

# Sample YAML 

Add a sample YAML file with `headService` with one head and one worker.

# Unit tests
It will be better to improve the test coverage of the unit tests.

* `Name`: If it's specified in HeadServiceSpec, override the default.

=> This overlap is not tested in `service_test.go`.

* `Namespace`: ignore user-provided namespace in HeadServiceSpec because we must enforce it to be the same namespace as the RayCluster.

=> This overlap is not tested in `service_test.go`.

* annotations field: The priority is Spec.HeadServiceAnnotations > custom HeadServiceSpec annotations.

=> The overlap is not tested in `service_test.go`.

* Selector field: Only use the keys from (2), don't add any new keys. Just update the values of the keys from (2) with the values from (1). Always ignore keys and values from (3).

=> Selector is not tested in `service_test.go`.

* Labels: field: Equal to Selector: + (3).

=> Check the relationship between Selector and Label.

* Ports field: There is already some logic in getPortsFromCluster to add default ports only if HeadGroupSpec.Template.Spec.Containers[index].Ports is empty. In this PR, we keep this logic and simply append all the ports from HeadServiceSpec. However, we can consider more complicated merging behavior in the future (or in this PR if necessary), like merging based on the key port.

* Type: If it's specified in HeadServiceSpec, override the default.

=> This is not tested in `service_test.go`.

architkulkarni commented on ray-operator/controllers/ray/raycluster_controller_test.go:
```diff
@@ -53,13 +53,41 @@ var _ = Context("Inside the default namespace", func() {
 		Spec: rayiov1alpha1.RayClusterSpec{
 			RayVersion:              "1.0",
 			EnableInTreeAutoscaling: &enableInTreeAutoscaling,
+			HeadServiceAnnotations: map[string]string{
```
Reverted all changes to `raycluster_controller_test.go`.  I will add all unit tests to `TestUserSpecifiedHeadService`.

architkulkarni commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -31,46 +31,124 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
+	default_type := cluster.Spec.HeadGroupSpec.ServiceType
+
+	defaultAppProtocol := DefaultServiceAppProtocol
+	ports_int := getServicePorts(cluster)
```
`ports_int` is a map from `str` to `int`, while `ports` is a list of `corev1.ServicePort`.  Added a code comment, but let me know if you have a better idea for the naming.

kevin85421 replied:
Got it.

architkulkarni commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -31,46 +31,124 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
+		// If the user has not specified a service type, use the cluster's service type
+		if headService.Spec.Type == "" {
+			headService.Spec.Type = default_type
+			log.Info("Using default service type for head service",
```
Great point, fixed.

architkulkarni commented on ray-operator/controllers/ray/common/service_test.go:
```diff
@@ -203,6 +204,78 @@ func TestGetServicePortsWithMetricsPort(t *testing.T) {
+	if headService.ObjectMeta.Labels[RayClusterLabelKey] != testRayClusterWithHeadService.ObjectMeta.Name {
+		t.Errorf("User cluster name label not found or incorrect value: key=%s, expected value=%s, actual value=%s", RayClusterLabelKey, testRayClusterWithHeadService.ObjectMeta.Name, headService.ObjectMeta.Labels[RayClusterLabelKey])
+	}
+	// Test merged annotations
```
Good point, done

kevin85421 commented on ray-operator/controllers/ray/common/service.go:
```diff
@@ -31,46 +31,124 @@ func BuildServiceForHeadPod(cluster rayiov1alpha1.RayCluster, labels map[string]
 	}
 
+	// Deep copy the selector to avoid modifying the original object
+	labels_for_service := make(map[string]string)
```
Oh, the `DeepCopy` function seems to be generated by the code-generator, so map will not have the function.

architkulkarni commented:
@kevin85421 I've addressed the comments, please take another look!
- Add sample YAML
- Revert "Add test in raycluster controller to execute control plane logic. Example: https://github.com/ray-project/kuberay/pull/1037" and move all tests to unit tests in `service_test.go`

kevin85421 reviewed (approved):
LGTM
